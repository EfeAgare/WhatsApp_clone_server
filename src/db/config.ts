import { Pool } from 'pg';
import 'dotenv/config';

 
export const environment = process.env.NODE_ENV || 'development';


export const dbConfig = {
  host: 'localhost',
  port: process.env.DB_PORT ? parseInt(process.env.DB_PORT) : 5433,
  user: process.env.BD_USER,
  password: '',
  database: process.env.DB_DATABASE,
};

if (environment === 'test') {
  dbConfig.database = process.env.DB_DATABASE_TEST
}

export let pool: Pool = new Pool(dbConfig)