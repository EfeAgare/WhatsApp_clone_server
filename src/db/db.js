import sql from 'sql-template-strings';
import 'dotenv/config';
import { pool } from './config';
import { resetDb as dbReset } from '../env';

export const dbSeeds = async () => {
  const sampleUsers = [
    {
      id: '1',
      phoneNumber: '08052467898',
      username: 'ray',
      aboutMe: "I'm getting better",
      picture: 'https://randomuser.me/api/portraits/thumb/lego/1.jpg',
    },
    {
      id: '2',
      username: 'ethan',
      phoneNumber: '09016787645',
      aboutMe: "I'm getting better",
      picture: 'https://randomuser.me/api/portraits/thumb/men/1.jpg',
    },
    {
      id: '3',
      username: 'bryan',
      phoneNumber: '09039678980',
      aboutMe: "I'm getting better",
      picture: 'https://randomuser.me/api/portraits/thumb/men/2.jpg',
    },
    {
      id: '4',
      username: 'avery',
      phoneNumber: '09039678547',
      aboutMe: "I'm getting better",
      picture: 'https://randomuser.me/api/portraits/thumb/women/1.jpg',
    },
    {
      id: '5',
      username: 'katie',
      phoneNumber: '09039678548',
      aboutMe: "I'm getting better",
      picture: 'https://randomuser.me/api/portraits/thumb/women/2.jpg',
    },
  ];

  for (const sampleUser of sampleUsers) {
    await pool.query(sql`
      INSERT INTO users(id, username, phoneNumber , picture, aboutMe)
      VALUES(${sampleUser.id}, ${sampleUser.username}, ${sampleUser.phoneNumber}, ${sampleUser.picture}, ${sampleUser.aboutMe})
    `);
  }

  await pool.query(sql`DELETE FROM chats`);

  const sampleChats = [
    {
      id: '1',
    },
    {
      id: '2',
    },
    {
      id: '3',
    },
    {
      id: '4',
    },
  ];

  for (const sampleChat of sampleChats) {
    await pool.query(sql`
      INSERT INTO chats(id)
      VALUES(${sampleChat.id})
    `);
  }


  await pool.query(sql`DELETE FROM chats_users`);

  const sampleChatsUsers = [
    {
      chat_id: '1',
      user_id: '1',
    },
    {
      chat_id: '1',
      user_id: '2',
    },
    {
      chat_id: '2',
      user_id: '1',
    },
    {
      chat_id: '2',
      user_id: '3',
    },
    {
      chat_id: '3',
      user_id: '1',
    },
    {
      chat_id: '3',
      user_id: '4',
    },
    {
      chat_id: '4',
      user_id: '1',
    },
    {
      chat_id: '4',
      user_id: '5',
    },
  ];

  for (const sampleChatUser of sampleChatsUsers) {
    await pool.query(sql`
      INSERT INTO chats_users(chat_id, user_id)
      VALUES(${sampleChatUser.chat_id}, ${sampleChatUser.user_id})
    `);
  }

  await pool.query(sql`DELETE FROM messages`);

  const baseTime = new Date('1 Jan 2019 GMT').getTime();

  const sampleMessages = [
    {
      id: '1',
      content: 'You on your way?',
      created_at: new Date(baseTime - 60 * 1000 * 1000),
      read: false,
      chat_id: '1',
      sender_user_id: '1',
    },
    {
      id: '2',
      content: "Hey, it's me",
      created_at: new Date(baseTime - 2 * 60 * 1000 * 1000),
      read: false,
      chat_id: '2',
      sender_user_id: '1',
    },
    {
      id: '3',
      content: 'I should buy a boat',
      created_at: new Date(baseTime - 24 * 60 * 1000 * 1000),
      chat_id: '3',
      read: false,
      sender_user_id: '1',
    },
    {
      id: '4',
      content: 'This is wicked good ice cream.',
      created_at: new Date(baseTime - 14 * 24 * 60 * 1000 * 1000),
      chat_id: '4',
      read: false,
      sender_user_id: '1',
    },
  ];

  for (const sampleMessage of sampleMessages) {
    await pool.query(sql`
      INSERT INTO messages(id, content, created_at, chat_id, sender_user_id, read)
      VALUES(${sampleMessage.id}, ${sampleMessage.content}, ${sampleMessage.created_at},
        ${sampleMessage.chat_id}, ${sampleMessage.sender_user_id}, ${sampleMessage.read})
    `);
  }

};
export const resetDb = async () => {
  dbSeeds;
};

if (dbReset) {
  resetDb();
}
