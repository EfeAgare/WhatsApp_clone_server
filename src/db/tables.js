import { pool } from './config';
import { dbSeeds } from './db';

const createChatsTable = `CREATE TABLE chats(
  id SERIAL PRIMARY KEY
);`;

const createUsersTable = `CREATE TABLE users(
  id SERIAL PRIMARY KEY,
  username VARCHAR (50) UNIQUE NOT NULL,
  name VARCHAR (50) NOT NULL,
  password VARCHAR (255) NOT NULL,
  picture VARCHAR (255) NOT NULL,
  aboutMe VARCHAR (100) NOT NULL
);`;

const createChatUsersTable = `CREATE TABLE chats_users(
  chat_id INTEGER NOT NULL REFERENCES chats(id) ON DELETE CASCADE,
  user_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE
);`;

const createMessagesTable = ` CREATE TABLE messages(
  id SERIAL PRIMARY KEY,
  content TEXT NOT NULL,
  created_at TIMESTAMP NOT NULL DEFAULT NOW(),
  chat_id INTEGER NOT NULL REFERENCES chats(id) ON DELETE CASCADE,
  sender_user_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE
);`;

const makeQuery = async (query) => {
  const db = await pool.connect();
  try {
    const res = await db.query(query);
    dbSeeds();
    db.release();
  } catch (error) {
    db.release();
    console.log(error.stack);
  }
};

makeQuery(
  `DROP TABLE IF EXISTS messages; DROP TABLE IF EXISTS chats_users; DROP TABLE IF EXISTS users; DROP TABLE IF EXISTS chats;${createChatsTable}${createUsersTable}${createChatUsersTable}${createMessagesTable}`
);
