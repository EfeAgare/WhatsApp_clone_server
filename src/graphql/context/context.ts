import { PubSub } from 'apollo-server-express';
import { User } from '../../db/db';
import { Response } from 'express';

export type MyContext = {
  pubsub: PubSub;
  currentUser: User;
  res: Response;
};
