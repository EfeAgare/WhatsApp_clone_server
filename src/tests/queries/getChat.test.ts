import { createTestClient } from 'apollo-server-testing';
import { ApolloServer, gql, PubSub } from 'apollo-server-express';
import schema from '../../graphql/index';
import { resetDb } from '../../db/db';
import sql from 'sql-template-strings';
import { pool } from '../../db/config';
import { MyContext } from '../../graphql/context/context';

describe('Query.chat', () => {
  afterAll(async () => await pool.end());

  it('should fetch specified chat', async () => {
    const { rows } = await pool.query(sql`SELECT * FROM users WHERE id = 2`);

    const currentUser = rows[0];

    const server = new ApolloServer({
      schema,
      context: async () => ({
        pubsub: new PubSub(),
        currentUser,
        db: await pool.connect(),
      }),
      // @ts-ignore
      formatResponse: (res: any, context: MyContext) => {
        context.db.release();
        return res;
      },
    });

    const { query } = createTestClient(server);

    const res = await query({
      variables: { chatId: '1' },
      query: gql`
        query GetChat($chatId: ID!) {
          chat(chatId: $chatId) {
            id
            name
            picture
            lastMessage {
              id
              content
              createdAt
            }
          }
        }
      `,
    });

    expect(res.data).toBeDefined();
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });
});
