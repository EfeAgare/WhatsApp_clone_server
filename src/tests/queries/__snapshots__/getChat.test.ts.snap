// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query.chat should fetch specified chat 1`] = `
Object {
  "chat": Object {
    "id": "1",
    "lastMessage": Object {
      "content": "You on your way?",
      "createdAt": "2020-04-30T22:10:42.179Z",
      "id": "1",
    },
    "name": "Ethan Gonzalez",
    "picture": "https://randomuser.me/api/portraits/thumb/men/1.jpg",
  },
}
`;
